ci: publish multi-arch image manifest lists

ngraef opened the pull request:
This change adds linux/arm64 binaries to the release. It also publishes an arm64 container image for all variants (standard, debug, rootless, static) and releases (dev, edge, latest).

The build and push process uses buildx to push the individual images by digest (i.e. untagged) and reference them in a single, tagged manifest list. This avoids cluttering Docker Hub's tag list with `<tag>-<arch>` tags.

Fixes #2233

srenatus commented on Makefile:
```diff
@@ -301,32 +312,69 @@ ci-build-windows: ensure-release-dir
+
+.PHONY: push-manifest-list-%
+push-manifest-list-%: ensure-executable-bin
+	$(DOCKER) buildx build \
```
Before, we built the images and _only pushed them_ in this target. Now, we're building them, aren't we? `docker buildx build` makes me think so. Would it make sense to also change the `image-quick` target to use `docker buildx ...`, and if not, what's the difference? 🤔

ngraef replied:
> Before, we built the images and only pushed them in this target. Now, we're building them, aren't we?

Yes, `buildx build --push` builds and pushes the images. There are now two different make targets for building images:
1. `image-quick` builds the image variants for a single platform, tags them, and loads them to the local docker image store. This target is used by the image smoke tests, which then `docker run` the local tagged images.
2. `push-manifest-list-%` builds and pushes untagged images for multiple platforms, then assembles the image references in a manifest list, tags it, and pushes it. This is used in the release process.

`buildx` does not support loading multi-platform manifests to the docker image store, so we can't use the same build→tag→push workflow that was previously used for single-platform images. It is possible to perform all the same steps `buildx build --push` does in separate commands/targets, but the single command is _significantly_ less complex. I also don't see a reason to break it up.

> Would it make sense to also change the image-quick target to use `docker buildx ...`

We could change `image-quick` to use `buildx build`, but we don't need any Buildx features here. It also requires the extra [`--load` flag](https://github.com/docker/buildx/blob/master/docs/reference/buildx_build.md#-load-the-single-platform-build-result-to-docker-images---load) to get the same result as `docker build`. If you think using `buildx` for everything is more important than the `docker build` simplicity, I can make that change. Either way, the image is built by BuildKit and runnable in the smoke tests.

srenatus replied:
> Either way, the image is built by BuildKit and runnable in the smoke tests.

So if I understand this correctly, it's more a matter of CLI commands used, not about what happens as an effect of those. That's cool, let's keep it. Thanks for taking the time to explain this! 👍

ngraef replied:
That's correct.

srenatus reviewed (commented):
Thank you for looking into this!

I've got a few questions, please bear with me, I've never used buildx before.

1. With the manifest list, can you still pull images by tag as before? I.e. `openpolicyagent/opa:0.36.1`, `openpolicyagent/opa:latest`, ``openpolicyagent/opa:edge`?
2. What does it look like on dockerhub? I'm sure you've got some example of the desired end state...
3. Our docker image smoke tests fail on the PR, it seems like we're missing something there: https://github.com/open-policy-agent/opa/runs/4879113973?check_suite_focus=true#step:4:59

Finally, do you have an idea how we could (smoke) test the produced binaries and images, preferable from github actions...? Are there any handy tools to emulate linux/arm64...?

ngraef commented:
> 1. With the manifest list, can you still pull images by tag as before? I.e. `openpolicyagent/opa:0.36.1`, `openpolicyagent/opa:latest`, `openpolicyagent/opa:edge`?

Yes! Clients continue to pull images the same way. `docker pull openpolicyagent/opa:0.36.1` defaults to pulling the image from the manifest list that matches the client architecture (`linux/amd64` in most cases). You can force pulling a specific architecture using the `--platform` option, like `docker pull --platform linux/arm64 openpolicyagent/opa:0.36.1`.

If you're curious to see what a manifest list contains, run `docker manifest inspect golang:1.17` or `docker buildx imagetools inspect golang:1.17`.

> 2. What does it look like on dockerhub? I'm sure you've got some example of the desired end state...

You've probably seen it on most of the "official" images, like the example from `golang` below. In our case, it will just be two architectures rather than the 10 shown.

<img width="1264" alt="Screen Shot 2022-01-20 at 2 58 16 AM" src="https://user-images.githubusercontent.com/1031317/150306279-d1857c59-86e6-48f0-99f8-d1742d1c1ab5.png">

> 3. Our docker image smoke tests fail on the PR, it seems like we're missing something there: https://github.com/open-policy-agent/opa/runs/4879113973?check_suite_focus=true#step:4:59

Already fixed it before seeing your review 😀 

> Finally, do you have an idea how we could (smoke) test the produced binaries and images, preferable from github actions...? Are there any handy tools to emulate linux/arm64...?

Docker has [setup-quemu-action](https://github.com/docker/setup-qemu-action) that is helpful. It allows you to build and run images for different architectures. We don't need it to build in this case because we're cross-compiling, but we could use it to test the built arm64 images.

srenatus commented:
Thanks for your replies there, it's very helpful to me. 😃 

> but we could use it to test the built arm64 images.

Would you mind looking into that, too? It would be great if we had some assurance in our pipeline before publishing arm64 images.

ngraef commented:
> > but we could use it to test the built arm64 images.
> 
> Would you mind looking into that, too? It would be great if we had some assurance in our pipeline before publishing arm64 images.

I added smoke tests for arm64 images. I also added the platform string to the `version` command output.

anderseknert commented:
Wow, this is an awesome contribution @ngraef 👏 

Just tried it and for the first time I can run OPA containerized on my M1 without the dreaded:
```
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
```
Nice! 👍

srenatus commented on Dockerfile:
```diff
@@ -6,20 +6,22 @@ ARG BASE
 
 FROM ${BASE}
 
+LABEL org.opencontainers.image.authors="Torin Sandall <torinsandall@gmail.com>"
```
Does this replace the `MAINTAINER` command completely...? 🤔

ngraef replied:
It does. [Docker says the `MAINTAINER` instruction is deprecated](https://docs.docker.com/engine/reference/builder/#maintainer-deprecated).

srenatus commented on Makefile:
```diff
@@ -301,107 +308,130 @@ ci-build-windows: ensure-release-dir
 
+.PHONY: ensure-executable-bin
+ensure-executable-bin:
+	find $(RELEASE_DIR) -type f ! -name "*.sha256" | xargs chmod +x
```
👏

srenatus commented on Makefile:
```diff
@@ -301,107 +308,130 @@ ci-build-windows: ensure-release-dir
-	$(DOCKER) run $(DOCKER_IMAGE):$(VERSION)-static version
+ci-image-smoke-test: ci-image-smoke-test-$(GOARCH)
+
+# % = arch
```
These comments are really helpful 👍

srenatus commented on version/version.go:
```diff
@@ -15,6 +15,9 @@ var Version = "0.37.0-dev"
 var GoVersion = runtime.Version()
 
+// Platform is the runtime OS and architecture of this OPA binary
+var Platform = runtime.GOOS + "/" + runtime.GOARCH
```
Does this really convey something interesting to the user? I'd imagine if you can run the binary _at all_, you know it's the binary for your platform; if it's not, you won't see this output.

srenatus replied:
Ignore me. Just noticed I can run `linux/arm64` using docker for mac. So this does convey something :+1:

ngraef replied:
That's exactly why I added it. I imagine it could also be helpful to maintainers when people file bug reports and paste the output of the `version` command.

srenatus reviewed (commented):
🎉 This looks great and I'm thankful for you both taking on this decent amount of work on our build processes, and taking the time to explain the changes.

I haven't played with it locally yet, I'm about to do that next.

srenatus commented on .github/workflows/pull-request.yaml:
```diff
@@ -34,6 +34,11 @@ jobs:
+          - os: linux
+            run: ubuntu-18.04
+            targets: ci-go-ci-build-linux ci-go-ci-build-linux-static
+            arch: arm64
```
Let's add the `matrix.arch` to the build name (line 27), it looks hard to discern right now:

![image](https://user-images.githubusercontent.com/870638/150757482-458945ae-9987-4944-82e3-7bed87cb191a.png)

srenatus replied:
We'll probably have to update the "required builds" then, but that's OK.

ngraef replied:
Updated. How's this?

<img width="221" alt="Screen Shot 2022-01-24 at 10 45 02 AM" src="https://user-images.githubusercontent.com/1031317/150826469-302ff0de-736f-46f7-abf8-994722f18fee.png">

srenatus replied:
Thanks

srenatus commented:
Could you please rebase and squash this, so we can get it merged?

bencooper222 commented:
I don't think this is happening in practice? This image was released 25 days ago and is x86-only.

https://hub.docker.com/layers/opa/openpolicyagent/opa/0.42.0/images/sha256-83ee526b78c6204ccc1fce0021d6c2eacfc2393e402a88260a994875448195ad?context=explore

anderseknert commented:
@bencooper222 you’ll need to use the static image https://hub.docker.com/layers/opa/openpolicyagent/opa/latest-static/images/sha256-b14bf57a46c2d73f117f8bab45788581cf2fddd87f16a888ba4e1d05d347b34d?context=explore

ngraef commented:
@bencooper222 This was partially reverted. See #4282 for more context.

timlaroche commented:
Hi all, just wondering if #4282 will be reversed as it seems that the issue in #4280 regarding https://github.com/bytecodealliance/wasmtime/issues/3183  has been closed as fixed? It would be nice to have the manifests for the non-static images also.

srenatus commented:
We still have problems testing it. There are not arm64 runners in Github, and we don't have the bandwidth to maintain our own arm-based runners on any other hardware. So we'd either have to ship something untested; or don't ship it at all

It could be worth retrying to test the images in Qemu. Last time I tried it, I ran into SIGILL exceptions.
